Initial API redesign

stylemistake opened the pull request:
Implemented the new API.

- Different code style for functions (`kebab-case`, which is more consistent with how commands are named in unix environment)
- Short aliases (#22)
- Accepts only full task names with prefixes (#18), mainly to leverage autocomplete in IDE, might also enable other interesting uses.
- Utility functions to traverse directories (they are simply logged better than standard pushd/popd)
- Custom task prefixes (to accommodate different code styles)

Documentation needs fixing.

Another interesting development vector is to try make this tool compatible with existing "standards" of task files.

- [Taskit](https://github.com/kjkuan/taskit) ([Taskfile](https://github.com/kjkuan/taskit/blob/master/Taskfile))
- [b5](https://pypi.org/project/b5/)

Compatibility with v0.x.x sadly was not preserved, but might fix this in the future.

stylemistake commented:
@SKalt would appreciate your opinion on all of this.

Runnerfile: https://github.com/stylemistake/runner/blob/next/runnerfile.sh
Alias list: https://github.com/stylemistake/runner/blob/next/src/runner.sh#L31-L44

SKalt commented on README.md:
```diff
@@ -215,7 +215,7 @@ All flags you pass after the task name are passed to your tasks.
 $ runner foo --production
 
 task_foo() {
```
Seems like this should be in kebab-case

SKalt commented:
I love the aliasing, indent, and documentation improvements.  I'm going to need to play around with the new API to figure out whether I agree with the changes for #18.  My current plan is to mess around with this tomorrow and Saturday (I'm interested in using [BATS](https://github.com/bats-core/bats-core) to extend the test suite).

stylemistake commented:
I'm questioning right now, whether we should still prefix task names with `task-`, because technically **anything** can be a task, e.g. if you call `@run-task ls` with new API, it will run the actual `ls` command.

Or, for example, running `ls` as a task from CLI:
```
[style@mia runner]$ bin/runner ls
[20:43:19.818] Starting 'ls'...
bin  completion  distrib  LICENSE.md  README.md  runnerfile.sh	src  test  VERSION
[20:43:19.823] Finished 'ls' after 2 ms
```

You don't even need a Runnerfile here to make pipelines, aka you should be able to write it all in one line.

And for CLI to know, which tasks are available to run, you can do something like this:
```
runner_export_tasks=(foo bar make-coffee)
```

stylemistake commented:
And regarding "writing pipeline in one line", I could not think of a better way other than using a [subarg spec](https://github.com/substack/subarg).

Browserify uses this spec to pass arguments to transforms and plugins (in our case, pass flags to tasks or programs).

Final API would look something like this:
```bash
# basic invocation
bin/runner task-a task-b
# -p means call `runner-run-tasks-parallel` with those tasks as arguments
bin/runner -p [ task-a task-b ]
# mix and match
bin/runner task-a [ task-b --flag ] -p [ task-c [ task-d --flag ] ]
```

stylemistake commented:
Zsh completions work differently than Bash, and need a separate implementation.

https://github.com/jez/bask/issues/1

SKalt commented:
Makes sense, and I missed the bit in `usage` about only bash being supported, so I'm going to drop that point.  Since the `task-` warts don't actually block runner calling an un-prefixed task*, I'm willing to sign off on this PR.


\* for example 
```bash
# runnerfile.sh
test-whether-task-prefix-required(){
  echo foo;
}

# on the cli
$  runner test-whether-task-prefix-required
[08:56:32.285] Starting 'test-whether-task-prefix-required'...
foo
[08:56:32.303] Finished 'test-whether-task-prefix-required' after 4 ms
```

stylemistake commented:
Small update on the progress:

Parallelism is now properly implemented! Tasks are now launched as workers which are launched by the master process and exchange information using named pipes.

Implementing inter-process communications using named pipes was a challenge, all while staying compatible with Bash v3, but I finally did it. I am still not sure, whether named pipes will work on Windows, but doing this was totally worth it. It is truly magical.

You can now define dependencies on tasks using function `@depends [<task> ...]` in the task body, which blocks execution and notifies the master about the dependency, which in turn spawns the worker for that task. Tasks with no dependencies run normally, and after they're finished, they notify the master to resolve dependencies, and after all dependencies are satisfied, `@depends` is unblocked and execution continues.

It is a really nice model in particular, because we don't even need to pre-build a graph to know what goes first, and everything is followed implicitly by the whole execution flow. Having a proper graph might help in the future, e.g. to find an optimal execution path for tasks before spawning them, but right now, it works as it is!

There are still a few minor issues left to solve; for example, it doesn't find and break cyclic dependencies, and enters a deadlock if it encounters one.

I also added a whole bunch of new source files, splitting various functional parts into logical modules, some of which are 100% dedicated for coping with lack of features in Bash v3 :smile: ([list.sh](https://github.com/stylemistake/runner/blob/next/src/list.sh) for example).

@SKalt: I reverted my decision on allowing the runner to accept full task names with prefixes, because it was really a pain in the ass to sort out once I started implementing the worker system.

Now, since I was heavily invested into compatibility with Bash v3, I made [colorize.sh](https://github.com/stylemistake/runner/blob/next/src/colorize.sh) Bash v3 compatible! Hooray!

There is also a particular feature I want to add before merging this to master and making a solid 1.0 release: `needs-rebuild` function. It's already implemented over here: https://github.com/stylemistake/runner/blob/next/src/misc.sh#L29-L68 but needs some more work and of course unit tests.

It works perfectly fine on its own, but it needs a tighter integration with the runner, so for example you could put it in the task body as a directive, e.g. `@needs-rebuild <output_file> -d <input_dir> -f <input_file> ...`, and if it doesn't need a rebuild, it will silently kill the worker subprocess, making it super easy to write GNU Make tier build pipelines!

And of course, here are some screenshots:

[Source code of the test](https://github.com/stylemistake/runner/blob/next/test/worker-model.sh)

![image](https://user-images.githubusercontent.com/1516236/54503447-eecdbb80-4937-11e9-9c9f-fb4297e6b3de.png)

<details>
<summary>Same thing, but with debug logging enabled:</summary>
<img src="https://user-images.githubusercontent.com/1516236/54503470-04db7c00-4938-11e9-9a29-46ef469d7201.png" alt="">
</details>

stylemistake commented:
After all this time, I have to close this branch. There are several factors that contributed to this:
- It breaks a lot of existing API, which is kind of bad and I thought about making it into a whole different project.
- Tasks that run in this new execution model can't work with stdin and color output is disabled in a lot of tools, which is a minor inconvenience and a side-effect of heavily tampering with stdio pipes.
- I no longer have a business need to continue development. It happened that this tool is, mainly used by sys-admins, which is reasonable because bash is used a lot. But when used in software development, it is very apparent that bash does not sit well with team members who don't know bash a lot, and much prefer tooling that is native to the language they develop on (e.g. C# or JavaScript).

For JavaScript, I have developed [Juke build](https://github.com/stylemistake/juke-build), which is sort of a prototype, but it is actively being used by [tgstation13](https://github.com/tgstation/tgstation], and offers a lot of key advantages over Runner.

A bit sad for Runner, but it is what it is.
